pyro/ops/tensor_utils.py

```diff
@@ -39,6 +39,76 @@ def block_diagonal(mat, block_size):
     return mat[..., ::B + 1, :, :]
 
 
+def periodic_repeat(tensor, size, dim):
+    """
+    Repeat a ``period``-sized tensor up to given ``size``. For example::
+
+        >>> x = torch.tensor([[1, 2, 3], [4, 5, 6]])
+        >>> periodic_repeat(x, 4, 0)
+        tensor([[1, 2, 3],
+                [4, 5, 6],
+                [1, 2, 3],
+                [4, 5, 6]])
+        >>> periodic_repeat(x, 4, 1)
+        tensor([[1, 2, 3, 1],
+                [4, 5, 6, 4]])
+
+    This is useful for computing static seasonality in time series models.
+
+    :param torch.Tensor tensor: A tensor of differences.
+    :param int size: Desired size of the result along dimension ``dim``.
+    :param int dim: The tensor dimension along which to repeat.
+    """
+    assert isinstance(size, int) and size >= 0
+    assert isinstance(dim, int)
+    if dim > 0:
+        dim -= tensor.dim()
+
+    period = tensor.size(dim)
+    repeats = [1] * tensor.dim()
+    repeats[dim] = (size + period - 1) // period
+    result = tensor.repeat(*repeats)
+    result = result[(Ellipsis, slice(None, size)) + (slice(None),) * (-1 - dim)]
+    return result
+
+
+def periodic_cumsum(tensor, period, dim):
+    """
+    Compute periodic cumsum along a given dimension. For example if dim=0::
+
+        for t in range(period):
+            assert result[t] == tensor[t]
+        for t in range(period, len(tensor)):
+            assert result[t] == tensor[t] + result[t - period]
+
+    This is useful for computing drifting seasonality in time series models.
+
+    :param torch.Tensor tensor: A tensor of differences.
+    :param int period: The period of repetition.
+    :param int dim: The tensor dimension along which to accumulate.
+    """
+    assert isinstance(period, int) and period > 0
+    assert isinstance(dim, int)
+    if dim > 0:
```

Comment:
Should this be `dim >= 0`, otherwise for dim = 0, I think the padding might be incorrect?

Reply:
Good catch, fixed.